src/picom.c

```diff
@@ -690,6 +690,47 @@ static struct managed_win *paint_preprocess(session_t *ps, bool *fade_running) {
 		if (was_painted && w->mode != mode_old) {
 			w->reg_ignore_valid = false;
 		}
+
+		// Transition
+		if (w->transition_direction && w->transition_time >= 0.0f && w->transition_time <= 1.0f) {
+			double transition = ps->o.transition_timing_function(w->transition_time);
+			
+			w->transition_time += ps->transition_intv; 
```

Comment:
Rather than calculator the time with refresh_rate, why not measuring the time directly? You can use `get_time_ms` which calls `clock_gettime`.

Reply:
Actually perhaps it's better to add an option `transition-step` in the same vein as `fade-in/out-step`, and use that as the increment.

Reply:
> Rather than calculator the time with refresh_rate, why not measuring the time directly? You can use `get_time_ms` which calls `clock_gettime`.

I tried calculating the delta time in each frame at first like how you did it in fading but that was not constant and transition was not as smooth as what i was looking for.

> Actually perhaps it's better to add an option `transition-step` in the same vein as `fade-in/out-step`, and use that as the increment.

Currently `transition-duration` will do `transition-step` work but with better explanation, i think 🤔. But maybe changing it to `transition-step` and change it's behavior to how fading works may make code base more constant.

---
`fade` and `transition` are very similar to each other because fading is actually a type of transition. I was thinking about having a config option of `transition-type` and have some types like `fade`, `transform` (current transition system, i didn't come up with a better name), `scale` (just an idea, didn't work on it yet), and can change transition type for each window, I think that will be cool, Any opinions?

Reply:
> Currently transition-duration will do transition-step work

Sorry I think I wasn't very clear. But what I meant was, instead of `w->transition_time += ps->transition_intv; `. You could add an option `transition-step`, and `w->transition_time += transition_step`. This is what fading in/out does. This way you don't need refresh_rate.

Reply:
wouldn't having a `transition-type` imply only one type of transition can be used at one time?

Reply:
> > Currently transition-duration will do transition-step work
> 
> Sorry I think I wasn't very clear. But what I meant was, instead of `w->transition_time += ps->transition_intv; `. You could add an option `transition-step`, and `w->transition_time += transition_step`. This is what fading in/out does. This way you don't need refresh_rate.

In that case transition will be faster in higher refresh rates. If are you ok with that, i will do it.

Reply:
Yeah, but that's what users are already doing for fading.

Reply:
Done